types/tx.go

```diff
@@ -102,6 +94,232 @@ func ToTxs(txs [][]byte) Txs {
+	// in the input size of the original list and the input size of the new list respectively.
+	// A 2 * M performance gain is possible but iterating all of the indexes simultaneously
+	// alongside the original list, but the multiple iterations were preferred to be more
+	// readable and maintainable.
+
+	// Sort a copy of the complete transaction slice so we can check for
+	// duplication. The copy is so we do not change the original ordering.
+	// Only the slices are copied, the transaction contents are shared.
+	allCopy := make([]Tx, len(t.all))
+	copy(allCopy, t.all)
+	sort.Sort(Txs(allCopy))
+
+	var size int64
+	for i, cur := range allCopy {
+		size += int64(len(cur))
+		if size > maxSizeBytes {
+			return fmt.Errorf("transaction data size %d exceeds maximum %d", size, maxSizeBytes)
+		}
+
+		// allCopy is sorted, so any duplicated data will be adjacent.
+		if i+1 < len(allCopy) && bytes.Equal(cur, allCopy[i+1]) {
+			return fmt.Errorf("found duplicate transaction with hash: %x", cur.Hash())
+		}
+	}
+
+	// create copies of each of the action-specific indexes so that order of the original
+	// indexes can be preserved.
+	addedCopy := make([]Tx, len(t.added))
+	copy(addedCopy, t.added)
+	sort.Sort(Txs(addedCopy))
+
+	removedCopy := make([]Tx, len(t.removed))
+	copy(removedCopy, t.removed)
+	sort.Sort(Txs(removedCopy))
+
+	unmodifiedCopy := make([]Tx, len(t.unmodified))
+	copy(unmodifiedCopy, t.unmodified)
+	sort.Sort(Txs(unmodifiedCopy))
+
+	// make a defensive copy of otxs so that the order of
+	// the caller's data is not altered.
+	otxsCopy := make([]Tx, len(otxs))
+	copy(otxsCopy, otxs)
+	sort.Sort(Txs(otxsCopy))
+
+	if ix, ok := containsAllTxs(otxsCopy, unmodifiedCopy); !ok {
+		return fmt.Errorf("new transaction incorrectly marked as removed, transaction hash: %x", unmodifiedCopy[ix].Hash())
+	}
+
+	if ix, ok := containsAllTxs(otxsCopy, removedCopy); !ok {
+		return fmt.Errorf("new transaction incorrectly marked as removed, transaction hash: %x", removedCopy[ix].Hash())
+	}
+	if ix, ok := containsNoneTxs(otxsCopy, addedCopy); !ok {
+		return fmt.Errorf("existing transaction incorrectly marked as added, transaction hash: %x", addedCopy[ix].Hash())
+	}
+	return nil
+}
+
+// containsNoneTxs checks that list a contains none of the transactions in list
+// b. If a match is found, the index in b of the matching transaction is returned.
+// Both lists must be sorted.
+func containsNoneTxs(a, b []Tx) (int, bool) {
+	aix, bix := 0, 0
+	for ; aix < len(a); aix++ {
+	LOOP:
+		for bix < len(b) {
+			switch bytes.Compare(b[bix], a[aix]) {
+			case 0:
+				return bix, false
+			case -1:
+				bix++
+				// we've reached the end of b, and the last value in b was
+				// smaller than the value under the iterator of a.
+				// a's values never get smaller, so we know there are no more matches
+				// in the list. We can terminate the iteration here.
+				if bix == len(b) {
+					return -1, true
+				}
+			case 1:
+				break LOOP
```

Comment:
Optional legibility suggestion:
```suggestion
nextA:
	for ; aix < len(a); aix++ {
		for ; bix < len(b); bix++ {
			switch bytes.Compare(b[bix], a[aix]) {
			case 0:
				return bix, false
			case -1:
				// we've reached the end of b, and the last value in b was
				// smaller than the value under the iterator of a.
				// a's values never get smaller, so we know there are no more matches
				// in the list. We can terminate the iteration here.
				if bix == len(b) {
					return -1, true
				}
			case 1:
				continue nextA
```

The C style ALLCAPS convention for labels isn't particularly typical—Effective Go recommends mixed-case for labels.

I'd also suggest not using very generic names like "LOOP", esp. when there are multiple loops. Perhaps "inner" vs "outer", or "nextA" vs "nextB" or something that will tell the reader more about the intent?

Reply:
nextA seems great.